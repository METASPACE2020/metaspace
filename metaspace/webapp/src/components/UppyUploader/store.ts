// based on https://uppy.io/docs/stores/#Implementing-Stores

import { ref } from 'vue'
import { State } from '@uppy/core'

type Listener = (prevState: State, nextState: State, patch: object) => void

export default () => {
  const state = ref<State | any>({})
  const listeners = new Set<Listener>()

  return {
    getState: () => state.value,
    setState: (patch: object) => {
      const prevState = state.value
      const nextState = { ...prevState, ...patch }

      state.value = nextState

      listeners.forEach((listener) => {
        listener(prevState, nextState, patch)
      })
    },
    subscribe: (listener: Listener) => {
      listeners.add(listener)
      return () => listeners.delete(listener)
    },
  }
}
