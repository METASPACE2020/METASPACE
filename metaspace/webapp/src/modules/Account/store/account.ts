import { Module } from 'vuex';
import { DialogType } from '../dialogs';
import router from '../../../router';

export interface AccountState {
  dialog: DialogType | null;
  showDialogAsPage: boolean;
}

const account = {
  namespaced: true,
  state: {
    dialog: null,
    showDialogAsPage: false
  },
  mutations: {
    showDialog(state: AccountState, dialog: DialogType) {
      state.dialog = dialog;
      state.showDialogAsPage = false;
    },
    showDialogAsPage(state: AccountState, dialog: DialogType) {
      state.dialog = dialog;
      state.showDialogAsPage = true;
    },
    hideDialog(state: AccountState, dialog: DialogType) {
      if (state.dialog === dialog) {
        state.dialog = null;
        if (state.showDialogAsPage) {
          router.push('/');
        }
      }
    }
  }
};

export default account as Module<AccountState, any>;
