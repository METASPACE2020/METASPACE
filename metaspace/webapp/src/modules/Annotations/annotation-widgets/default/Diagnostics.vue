<template>
<div>
    <el-row id="scores-table">
        <div class="msm-score-calc">
            MSM score =
            <span>{{ annotation.msmScore.toFixed(3) }}</span> =
            <span>{{ annotation.rhoSpatial.toFixed(3) }}</span>
            (&rho;<sub>spatial</sub>) &times;
            <span>{{ annotation.rhoSpectral.toFixed(3) }}</span>
            (&rho;<sub>spectral</sub>) &times;
            <span>{{ annotation.rhoChaos.toFixed(3) }}</span>
            (&rho;<sub>chaos</sub>)
        </div>
        <div v-if="showOffSample">
            <el-popover trigger="hover" :open-delay="100">
                Image analysis gave an off-sample probability of {{formattedOffSampleProb}}.
                <span slot="reference" :class="annotation.offSample ? 'off-sample-tag' : 'on-sample-tag'">
                    {{annotation.offSample ? 'Off-sample' : 'On-sample'}}
                </span>
            </el-popover>
        </div>
    </el-row>
    <el-row id="isotope-images-container">
        <el-col :xs="24" :sm="12" :md="12" :lg="6"
                v-for="(img, idx) in sortedIsotopeImages"
                :key="annotation.id + idx">
            <div class="small-peak-image">
            {{ img.mz.toFixed(4) }}<br/>
                <image-loader :src="img.url"
                              :colormap="colormap"
                              :imageFitParams="{areaMinHeight: 50, areaHeight: 250}"
                              v-bind="imageLoaderSettings"
                              style="overflow: hidden"
                              :minIntensity="img.minIntensity"
                              :maxIntensity="img.maxIntensity"
                              showPixelIntensity
                />
            </div>
        </el-col>
    </el-row>
    <el-row id="isotope-plot-container">
        <isotope-pattern-plot :data="peakChartData"
                              :isotopeColors="Array(annotation.isotopeImages.length).fill(sampleIsotopeColor)"
                              :theorColor="theorIsotopeColor">
        </isotope-pattern-plot>
    </el-row>
    <el-row>
        <plot-legend :items="isotopeLegendItems">
        </plot-legend>
    </el-row>
</div>
</template>

<script lang="ts">
import Vue from 'vue';
import { Component, Prop } from 'vue-property-decorator';

import ImageLoader from '../../../../components/ImageLoader.vue';
import PlotLegend from '../PlotLegend.vue';
import IsotopePatternPlot from '../IsotopePatternPlot.vue';
import {sortBy} from 'lodash-es';
import config from '../../../../config'

@Component({
    name: 'diagnostics',
    components: {
        ImageLoader,
        PlotLegend,
        IsotopePatternPlot
    }
})
export default class Diagnostics extends Vue {
    @Prop()
    annotation: any
    @Prop()
    peakChartData: any
    @Prop()
    colormap: any
    @Prop()
    imageLoaderSettings: any

    sampleIsotopeColor: string = 'red'
    theorIsotopeColor: string = 'blue'

    get isotopeLegendItems(): any[] {
        return this.annotation ? [{name: 'Sample', color: this.sampleIsotopeColor, opacity: 1},
                                  {name: 'Theoretical', color: this.theorIsotopeColor, opacity: 0.6}]
                               : [];
    }

    get sortedIsotopeImages(): any[] {
        // Usually isotope images are pre-sorted by the server, but it's not an explicit guarantee of the API
        return sortBy(this.annotation.isotopeImages, img => img.mz)
          .filter(img => img.mz > 0);
    }

    get showOffSample(): boolean {
        return config.features.off_sample && this.annotation.offSample != null;
    }

    get formattedOffSampleProb(): string {
        if (this.annotation.offSampleProb < 0.1) {
            return 'less than 10%';
        } else if (this.annotation.offSampleProb > 0.9) {
            return 'greater than 90%';
        } else {
            return (+this.annotation.offSampleProb * 100).toFixed(0) + '%'
        }
    }
}
</script>

<style lang="scss" scoped>
@import "~element-ui/packages/theme-chalk/src/common/var";

#scores-table {
    display: flex;
    border-collapse: collapse;
    border: 1px solid lightblue;
    font-size: 16px;
    padding: 3px 10px;
    align-items: center;
}

.msm-score-calc {
    text-align: center;
    flex-grow: 1;
    justify-content: center;
}

.msm-score-calc > span {
    color: blue;
}

#isotope-images-container {
    margin: 10px auto;
    text-align: center;
    font-size: 13px;
}

.small-peak-image {
    font-size: 1rem;
    vertical-align: top;
    padding: 0 5px 0 5px;
    text-align: center;
    flex: 0 1 260px;
    box-sizing: border-box;

    @media (max-width: 768px) {
        flex-basis: 100%;
    }
}

    .off-sample-tag {
        margin-left: 10px;
        color: white;
        padding: 0 3px;
        background: $--color-warning;
        border-radius: 3px;
    }
    .on-sample-tag {
        margin-left: 10px;
        color: white;
        padding: 0 3px;
        background: $--color-success;
        border-radius: 3px;
    }
</style>
