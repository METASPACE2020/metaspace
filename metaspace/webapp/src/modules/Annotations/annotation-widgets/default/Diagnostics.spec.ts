import { flushPromises, mount } from '@vue/test-utils'
import router from '../../../../router'
import Diagnostics from './Diagnostics.vue'
import store from '../../../../store/index'
import { vi } from 'vitest'
import { nextTick, ref } from 'vue'
import * as reportErrorModule from '../../../../lib/reportError'

vi.mock('../../../../lib/ionImageRendering')
vi.mock('upng-js', () => ({
  Image: vi.fn(() => 'http://placebacon.com/200/300'),
}))

const peakChartData = JSON.stringify({
  mz_grid: { min_mz: 99.9, max_mz: 101.1 },
  theor: {
    centroid_mzs: [100, 101],
    mzs: [99.9, 100, 100.1, 100.9, 101, 101.1],
    ints: [0, 1, 0, 0, 0.5, 0],
  },
  ppm: 3,
  sampleData: {
    mzs: [100, 101],
    ints: [1, 0.6],
  },
})

const metrics = {
  chaos: 0.9976041666666666,
  spatial: 0.005921380836044965,
  spectral: 0.7906196869421936,
}
const metricsJson = JSON.stringify(metrics)
const baseAnnotation = {
  id: '2019-10-15_17h02m31s_7277991',
  ion: 'C19H18N2O7S2-H+',
  ionFormula: 'C19H19N2O7S2',
  mz: 449.0482267014823,
  fdrLevel: 0.1,
  msmScore: 0.657486,
  rhoSpatial: 0.686362,
  rhoSpectral: 0.959122,
  rhoChaos: 0.998756,
  peakChartData,
  possibleCompounds: [{ name: 'C.I. Food Red 6', information: [{ databaseId: 'HMDB0032738' }] }],
  dataset: {
    id: '2019-02-12_15h55m06s',
    name: 'Untreated_3_434',
    scoringModel: null,
    configJson: JSON.stringify({
      fdr: { scoring_model: null },
    }),
  },
  databaseDetails: {
    id: 24,
  },
  offSample: false,
  offSampleProb: 0.03,
  isobars: [
    {
      ion: 'C8H10S2-H-',
      ionFormula: 'C8H9S2',
      peakNs: [
        [1, 1],
        [3, 3],
      ],
      shouldWarn: true,
    },
  ],
  isotopeImages: [
    {
      mz: 100,
      url: '/fs/iso_images/img1',
      minIntensity: 0,
      maxIntensity: 100,
      totalIntensity: 5000,
    },
    {
      mz: 101,
      url: '/fs/iso_images/img2',
      minIntensity: 100,
      maxIntensity: 200,
      totalIntensity: 10000,
    },
  ],
  metricsJson,
}
const isobarAnnotation = {
  ...baseAnnotation,
  id: '15_17h02m31s_7277992',
  ion: 'C8H10S2-H-',
  ionFormula: 'C8H9S2',
  offSample: true,
  offSampleProb: 0.87654,
}

const props = {
  annotation: baseAnnotation,
  colormap: 'Viridis',
  imageLoaderSettings: {
    annotImageOpacity: 1,
    opacityMode: 'constant',
    imagePosition: { zoom: 1, xOffset: 0, yOffset: 0 },
    opticalSrc: null,
    opticalTransform: null,
    pixelAspectRatio: 1,
  },
}

const stubs = ['ImageLoader']

const W = 200
const H = 300

describe('Diagnostics', () => {
  beforeEach(() => {
    // Set HTMLElements to have non-zero dimensions
    vi.spyOn(HTMLElement.prototype, 'getBoundingClientRect').mockImplementation(() =>
      // @ts-ignore
      ({ left: 200, right: 200 + W, top: 100, bottom: 100 + H, width: W, height: H })
    )
    vi.spyOn(HTMLElement.prototype, 'clientWidth', 'get').mockImplementation(() => W)
    vi.spyOn(HTMLElement.prototype, 'clientHeight', 'get').mockImplementation(() => H)
  })

  it('should match snapshot (no isobars)', async () => {
    vi.mock('@vue/apollo-composable', () => ({
      useQuery: vi.fn(() => ({
        result: ref({
          allAnnotations: [],
          annotation: ({ id }: { id: string }) => ({ ...baseAnnotation, id, peakChartData }),
        }),
        loading: ref(false),
      })),
    }))

    const propsData = {
      ...props,
      annotation: {
        ...baseAnnotation,
        isobars: [],
      },
    }
    const wrapper = mount(Diagnostics, {
      global: {
        plugins: [store, router],
      },
      stubs,
      propsData,
    })
    await flushPromises()
    await nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should match snapshot (with isobar)', async () => {
    vi.mock('@vue/apollo-composable', () => ({
      useQuery: vi.fn(() => ({
        result: ref({
          allAnnotations: [isobarAnnotation],
          annotation: (_: any, { id }: { id: string }) => ({
            ...(id === isobarAnnotation.id ? isobarAnnotation : baseAnnotation),
            id,
            peakChartData,
          }),
        }),
        loading: ref(false),
      })),
    }))
    const reportErrorFunc = vi.spyOn(reportErrorModule, 'default')

    const wrapper = mount(Diagnostics, {
      global: {
        plugins: [store, router],
      },
      stubs,
      propsData: props,
    })

    await flushPromises()
    await nextTick()

    wrapper.vm.comparisonIonFormula = isobarAnnotation.ionFormula

    await flushPromises()
    await nextTick()

    expect(reportErrorFunc).not.toBeCalled() // "Inconsistent annotations" warning should fail the test
    expect(wrapper.html()).toMatchSnapshot()
  })
})
