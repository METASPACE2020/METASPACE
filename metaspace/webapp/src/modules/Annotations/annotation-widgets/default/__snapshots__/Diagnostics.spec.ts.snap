// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Diagnostics should match snapshot (no isobars) 1`] = `
<div mock-v-loading="0">
  <!---->
  <!---->
  <!---->
  <div class="">
    <div class="el-row" id="scores-table">
      <div class="msm-score-calc">
        MSM score =
        <span>0.657</span> =
        <span>0.686</span> (ρ
        <sub>spatial</sub>) ×
        <span>0.959</span> (ρ
        <sub>spectral</sub>) ×
        <span>0.999</span> (ρ
        <sub>chaos</sub>)
      </div>
      <div>
        <mock-el-popover trigger="hover" open-delay="100">
          <div slot-key="default">
            Image analysis gave an off-sample probability of less than 10%.
          </div>
          <div slot-key="reference"><span class="on-sample-tag">
                On-sample
            </span></div>
        </mock-el-popover>
      </div>
    </div>
    <div class="isotope-images-container el-row">
      <div class="el-col el-col-24 el-col-xs-24 el-col-sm-12 el-col-md-12 el-col-lg-6">
        <div class="small-peak-image">
          100.0000
          <br>
          <imageloader-stub src="/fs/iso_images/img1" colormap="Viridis" imagefitparams="[object Object]" minintensity="0" maxintensity="100" showpixelintensity="" annotimageopacity="1" opacitymode="constant" imageposition="[object Object]" pixelaspectratio="1"
            style="overflow: hidden;"></imageloader-stub>
        </div>
      </div>
      <div class="el-col el-col-24 el-col-xs-24 el-col-sm-12 el-col-md-12 el-col-lg-6">
        <div class="small-peak-image">
          101.0000
          <br>
          <imageloader-stub src="/fs/iso_images/img2" colormap="Viridis" imagefitparams="[object Object]" minintensity="100" maxintensity="200" showpixelintensity="" annotimageopacity="1" opacitymode="constant" imageposition="[object Object]" pixelaspectratio="1"
            style="overflow: hidden;"></imageloader-stub>
        </div>
      </div>
    </div>
  </div>
  <!---->
  <!---->
  <div>
    <div class="el-row" id="isotope-plot-container">
      <div class="peak-chart">
        <div style="height: 300px;">
          <svg width="0" height="0">
            <defs>
              <pattern id="pattern-stripe" width="6" height="6" patternUnits="userSpaceOnUse" patternTransform="rotate(45)">
                <rect width="3" height="6" transform="translate(0,0)" fill="white"></rect>
              </pattern>
              <mask id="mask-stripe">
                <rect x="0" y="0" width="100%" height="100%" fill="url(#pattern-stripe)"></rect>
              </mask>
            </defs>
            <g transform="translate(40, 10)">
              <g transform="translate(0, -60)" fill="none" font-size="10" font-family="sans-serif" text-anchor="middle">
                <path class="domain" stroke="currentColor" d="M0.5,6V0.5H-79.5V6"></path>
                <g class="tick" opacity="1" transform="translate(-6.166666666666351,0)">
                  <line stroke="currentColor" y2="6"></line>
                  <text fill="currentColor" y="9" dy="0.71em">100.0</text>
                </g>
                <g class="tick" opacity="1" transform="translate(-19.5,0)">
                  <line stroke="currentColor" y2="6"></line>
                  <text fill="currentColor" y="9" dy="0.71em">100.2</text>
                </g>
                <g class="tick" opacity="1" transform="translate(-32.83333333333365,0)">
                  <line stroke="currentColor" y2="6"></line>
                  <text fill="currentColor" y="9" dy="0.71em">100.4</text>
                </g>
                <g class="tick" opacity="1" transform="translate(-46.16666666666635,0)">
                  <line stroke="currentColor" y2="6"></line>
                  <text fill="currentColor" y="9" dy="0.71em">100.6</text>
                </g>
                <g class="tick" opacity="1" transform="translate(-59.5,0)">
                  <line stroke="currentColor" y2="6"></line>
                  <text fill="currentColor" y="9" dy="0.71em">100.8</text>
                </g>
                <g class="tick" opacity="1" transform="translate(-72.83333333333366,0)">
                  <line stroke="currentColor" y2="6"></line>
                  <text fill="currentColor" y="9" dy="0.71em">101.0</text>
                </g>
              </g>
              <g fill="none" font-size="10" font-family="sans-serif" text-anchor="end">
                <path class="domain" stroke="currentColor" d="M-6,-59.5H0.5V0.5H-6"></path>
                <g class="tick" opacity="1" transform="translate(0,-59.5)">
                  <line stroke="currentColor" x2="-6"></line>
                  <text fill="currentColor" x="-11" dy="0.32em">0</text>
                </g>
                <g class="tick" opacity="1" transform="translate(0,-47.5)">
                  <line stroke="currentColor" x2="-6"></line>
                  <text fill="currentColor" x="-11" dy="0.32em">20</text>
                </g>
                <g class="tick" opacity="1" transform="translate(0,-35.5)">
                  <line stroke="currentColor" x2="-6"></line>
                  <text fill="currentColor" x="-11" dy="0.32em">40</text>
                </g>
                <g class="tick" opacity="1" transform="translate(0,-23.5)">
                  <line stroke="currentColor" x2="-6"></line>
                  <text fill="currentColor" x="-11" dy="0.32em">60</text>
                </g>
                <g class="tick" opacity="1" transform="translate(0,-11.5)">
                  <line stroke="currentColor" x2="-6"></line>
                  <text fill="currentColor" x="-11" dy="0.32em">80</text>
                </g>
                <g class="tick" opacity="1" transform="translate(0,0.5)">
                  <line stroke="currentColor" x2="-6"></line>
                  <text fill="currentColor" x="-11" dy="0.32em">100</text>
                </g>
              </g>
              <text style="text-anchor: middle; text-decoration-line: underline; text-decoration-style: dashed;" transform="translate(-30, -30) rotate(-90)">Relative intensity</text>
              <text style="text-anchor: middle;" transform="translate(-40, -30) ">m/z</text>
              <g class="sample-graph soloSample">
                <g>
                  <circle cx="-6.666666666666351" cy="0" r="4"></circle>
                  <circle cx="-73.33333333333366" cy="-24" r="4"></circle>
                </g>
                <g>
                  <line x1="-6.666666666666351" x2="-6.666666666666351" y1="0" y2="-60"></line>
                  <line x1="-73.33333333333366" x2="-73.33333333333366" y1="-24" y2="-60"></line>
                </g>
                <g>
                  <rect width="-0.03999999999943338" height="-60" x="-6.6466666666666345" y="0"></rect>
                  <rect width="-0.04040000000031796" height="-36" x="-73.3131333333335" y="-24"></rect>
                </g>
              </g>
              <g class="theor-graph soloTheor">
                <g>
                  <path d="M0,-60L-6.666666666666351,-59.4L-13.333333333332702,-60L-66.6666666666673,-60L-73.33333333333366,-59.7L-80,-60"></path>
                </g>
              </g>
              <g fill="none" pointer-events="all">
                <rect class="overlay" pointer-events="all" cursor="crosshair" x="0" y="0" width="-80" height="-60"></rect>
                <rect class="selection" cursor="move" fill="#777" fill-opacity="0.3" stroke="#fff" shape-rendering="crispEdges" style="display: none;"></rect>
                <rect class="handle handle--w" cursor="ew-resize" style="display: none;"></rect>
                <rect class="handle handle--e" cursor="ew-resize" style="display: none;"></rect>
              </g>
            </g>
          </svg>
        </div>
        <div class="plot-legend">
          <div class="legend-item">
            <svg width="64" height="32">
              <g>
                <g class="theor-graph soloTheor">
                  <g>
                    <path d="M0,31L2,31L4,31L6,31L8,31L10,30.7L12,30.4L14,29.8L16,28.3L18,26.2L20,22.9L22,18.4L24,13.3L26,7.900000000000002L28,3.6999999999999993L30,1L32,1L34,3.400000000000002L36,7.900000000000002L38,13L40,18.4L42,22.6L44,26.2L46,28.3L48,29.5L50,30.4L52,30.7L54,31L56,31L58,31L60,31L62,31"></path>
                  </g>
                </g>
              </g>
            </svg> <span class="legend-item-name">Theoretical</span></div>
          <div class="legend-item">
            <svg width="64" height="32">
              <g>
                <g class="sample-graph soloSample">
                  <g>
                    <circle cx="32" cy="4" r="4"></circle>
                  </g>
                  <g>
                    <line x1="32" x2="32" y1="4" y2="32"></line>
                  </g>
                  <g>
                    <rect width="48" height="28" x="8" y="4"></rect>
                  </g>
                </g>
              </g>
            </svg> <span class="legend-item-name">Sample</span></div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Diagnostics should match snapshot (with isobar) 1`] = `
<div mock-v-loading="0">
  <div role="alert" class="el-alert isobar-alert el-alert--warning is-light">
    <i class="el-alert__icon el-icon-warning is-big"></i>
    <div class="el-alert__content">
      <!---->
      <p class="el-alert__description">
        Another ion was annotated that is isobaric to the selected annotation.
        <!---->
        Select an isobaric annotation below to compare against.
      </p>
      <!---->
      <i class="el-alert__closebtn el-icon-close" style="display: none;"></i>
    </div>
  </div>
  <div class="compare-container">
    Compare selected annotation to:
    <mock-el-select placeholder="None" clearable="" class="compare-select" value="C8H9S2">
      <!---->
      <mock-el-option value="C8H9S2" label="[M+0.0000]: [C8H10S2 - H]⁻">[M+0.0000]: [C
        <sub>8</sub>H
        <sub>10</sub>S
        <sub>2</sub> - H]⁻</mock-el-option>
    </mock-el-select>
  </div>
  <div class="ref-annotation-header">
    <mock-el-popover trigger="hover" placement="top" opendelay="100">
      <div slot-key="default">
        <div><span>Candidate molecules (1):</span>
          <ul>
            <li>
              C.I. Food Red 6
            </li>
            <!---->
          </ul>
          <!---->
          <!---->
        </div>
      </div>
      <div slot-key="reference"><span class="annotation-ion">[C<sub>19</sub>H<sub>18</sub>N<sub>2</sub>O<sub>7</sub>S<sub>2</sub> - H]⁺</span></div>
    </mock-el-popover> <span style="padding-left: 5px;">(Selected annotation)</span></div>
  <div class="ref-annotation-container">
    <div class="el-row" id="scores-table">
      <div class="msm-score-calc">
        MSM score =
        <span>0.657</span> =
        <span>0.686</span> (ρ
        <sub>spatial</sub>) ×
        <span>0.959</span> (ρ
        <sub>spectral</sub>) ×
        <span>0.999</span> (ρ
        <sub>chaos</sub>)
      </div>
      <div>
        <mock-el-popover trigger="hover" open-delay="100">
          <div slot-key="default">
            Image analysis gave an off-sample probability of less than 10%.
          </div>
          <div slot-key="reference"><span class="on-sample-tag">
                On-sample
            </span></div>
        </mock-el-popover>
      </div>
    </div>
    <div class="isotope-images-container el-row">
      <div class="el-col el-col-24 el-col-xs-24 el-col-sm-12 el-col-md-12 el-col-lg-6">
        <div class="small-peak-image">
          100.0000
          <br>
          <imageloader-stub src="/fs/iso_images/img1" colormap="Viridis" imagefitparams="[object Object]" minintensity="0" maxintensity="100" showpixelintensity="" annotimageopacity="1" opacitymode="constant" imageposition="[object Object]" pixelaspectratio="1"
            style="overflow: hidden;"></imageloader-stub>
        </div>
      </div>
      <div class="el-col el-col-24 el-col-xs-24 el-col-sm-12 el-col-md-12 el-col-lg-6">
        <div class="small-peak-image">
          101.0000
          <br>
          <imageloader-stub src="/fs/iso_images/img2" colormap="Viridis" imagefitparams="[object Object]" minintensity="100" maxintensity="200" showpixelintensity="" annotimageopacity="1" opacitymode="constant" imageposition="[object Object]" pixelaspectratio="1"
            style="overflow: hidden;"></imageloader-stub>
        </div>
      </div>
    </div>
  </div>
  <div class="comp-annotation-header">
    <!----><span><mock-el-popover trigger="hover" placement="top" opendelay="100"><div slot-key="default"><div><span>Candidate molecules (1):</span>
    <ul>
      <li>
        C.I. Food Red 6
      </li>
      <!---->
    </ul>
    <!---->
    <!---->
  </div>
</div>
<div slot-key="reference"><span class="annotation-ion">[C<sub>8</sub>H<sub>10</sub>S<sub>2</sub> - H]⁻</span></div>
</mock-el-popover>
<!---->
</span>
</div>
<div class="comp-annotation-container">
  <div class="el-row" id="scores-table">
    <div class="msm-score-calc">
      MSM score =
      <span>0.657</span> =
      <span>0.686</span> (ρ
      <sub>spatial</sub>) ×
      <span>0.959</span> (ρ
      <sub>spectral</sub>) ×
      <span>0.999</span> (ρ
      <sub>chaos</sub>)
    </div>
    <div>
      <mock-el-popover trigger="hover" open-delay="100">
        <div slot-key="default">
          Image analysis gave an off-sample probability of 88%.
        </div>
        <div slot-key="reference"><span class="off-sample-tag">
                Off-sample
            </span></div>
      </mock-el-popover>
    </div>
  </div>
  <div class="isotope-images-container el-row">
    <div class="el-col el-col-24 el-col-xs-24 el-col-sm-12 el-col-md-12 el-col-lg-6">
      <div class="small-peak-image">
        100.0000
        <br>
        <imageloader-stub src="/fs/iso_images/img1" colormap="Viridis" imagefitparams="[object Object]" minintensity="0" maxintensity="100" showpixelintensity="" annotimageopacity="1" opacitymode="constant" imageposition="[object Object]" pixelaspectratio="1"
          style="overflow: hidden;"></imageloader-stub>
      </div>
    </div>
    <div class="el-col el-col-24 el-col-xs-24 el-col-sm-12 el-col-md-12 el-col-lg-6">
      <div class="small-peak-image">
        101.0000
        <br>
        <imageloader-stub src="/fs/iso_images/img2" colormap="Viridis" imagefitparams="[object Object]" minintensity="100" maxintensity="200" showpixelintensity="" annotimageopacity="1" opacitymode="constant" imageposition="[object Object]" pixelaspectratio="1"
          style="overflow: hidden;"></imageloader-stub>
      </div>
    </div>
  </div>
</div>
<div>
  <div class="el-row" id="isotope-plot-container">
    <div class="peak-chart">
      <div style="height: 300px;">
        <svg width="0" height="0">
          <defs>
            <pattern id="pattern-stripe" width="6" height="6" patternUnits="userSpaceOnUse" patternTransform="rotate(45)">
              <rect width="3" height="6" transform="translate(0,0)" fill="white"></rect>
            </pattern>
            <mask id="mask-stripe">
              <rect x="0" y="0" width="100%" height="100%" fill="url(#pattern-stripe)"></rect>
            </mask>
          </defs>
          <g transform="translate(40, 10)">
            <g transform="translate(0, -60)" fill="none" font-size="10" font-family="sans-serif" text-anchor="middle">
              <path class="domain" stroke="currentColor" d="M0.5,6V0.5H-79.5V6"></path>
              <g class="tick" opacity="1" transform="translate(-6.166666666666351,0)">
                <line stroke="currentColor" y2="6"></line>
                <text fill="currentColor" y="9" dy="0.71em">100.0</text>
              </g>
              <g class="tick" opacity="1" transform="translate(-19.5,0)">
                <line stroke="currentColor" y2="6"></line>
                <text fill="currentColor" y="9" dy="0.71em">100.2</text>
              </g>
              <g class="tick" opacity="1" transform="translate(-32.83333333333365,0)">
                <line stroke="currentColor" y2="6"></line>
                <text fill="currentColor" y="9" dy="0.71em">100.4</text>
              </g>
              <g class="tick" opacity="1" transform="translate(-46.16666666666635,0)">
                <line stroke="currentColor" y2="6"></line>
                <text fill="currentColor" y="9" dy="0.71em">100.6</text>
              </g>
              <g class="tick" opacity="1" transform="translate(-59.5,0)">
                <line stroke="currentColor" y2="6"></line>
                <text fill="currentColor" y="9" dy="0.71em">100.8</text>
              </g>
              <g class="tick" opacity="1" transform="translate(-72.83333333333366,0)">
                <line stroke="currentColor" y2="6"></line>
                <text fill="currentColor" y="9" dy="0.71em">101.0</text>
              </g>
            </g>
            <g fill="none" font-size="10" font-family="sans-serif" text-anchor="end">
              <path class="domain" stroke="currentColor" d="M-6,-59.5H0.5V0.5H-6"></path>
              <g class="tick" opacity="1" transform="translate(0,-59.5)">
                <line stroke="currentColor" x2="-6"></line>
                <text fill="currentColor" x="-11" dy="0.32em">0</text>
              </g>
              <g class="tick" opacity="1" transform="translate(0,-47.5)">
                <line stroke="currentColor" x2="-6"></line>
                <text fill="currentColor" x="-11" dy="0.32em">20</text>
              </g>
              <g class="tick" opacity="1" transform="translate(0,-35.5)">
                <line stroke="currentColor" x2="-6"></line>
                <text fill="currentColor" x="-11" dy="0.32em">40</text>
              </g>
              <g class="tick" opacity="1" transform="translate(0,-23.5)">
                <line stroke="currentColor" x2="-6"></line>
                <text fill="currentColor" x="-11" dy="0.32em">60</text>
              </g>
              <g class="tick" opacity="1" transform="translate(0,-11.5)">
                <line stroke="currentColor" x2="-6"></line>
                <text fill="currentColor" x="-11" dy="0.32em">80</text>
              </g>
              <g class="tick" opacity="1" transform="translate(0,0.5)">
                <line stroke="currentColor" x2="-6"></line>
                <text fill="currentColor" x="-11" dy="0.32em">100</text>
              </g>
            </g>
            <text style="text-anchor: middle; text-decoration-line: underline; text-decoration-style: dashed;" transform="translate(-30, -30) rotate(-90)">Relative intensity</text>
            <text style="text-anchor: middle;" transform="translate(-40, -30) ">m/z</text>
            <g class="sample-graph soloSample">
              <g>
                <circle cx="-6.666666666666351" cy="0" r="4"></circle>
                <circle cx="-73.33333333333366" cy="-24" r="4"></circle>
              </g>
              <g>
                <line x1="-6.666666666666351" x2="-6.666666666666351" y1="0" y2="-60"></line>
                <line x1="-73.33333333333366" x2="-73.33333333333366" y1="-24" y2="-60"></line>
              </g>
              <g>
                <rect width="-0.03999999999943338" height="-60" x="-6.6466666666666345" y="0"></rect>
                <rect width="-0.04040000000031796" height="-36" x="-73.3131333333335" y="-24"></rect>
              </g>
            </g>
            <g class="theor-graph soloTheor">
              <g>
                <path d="M0,-60L-6.666666666666351,-59.4L-13.333333333332702,-60L-66.6666666666673,-60L-73.33333333333366,-59.7L-80,-60"></path>
              </g>
            </g>
            <g fill="none" pointer-events="all">
              <rect class="overlay" pointer-events="all" cursor="crosshair" x="0" y="0" width="-80" height="-60"></rect>
              <rect class="selection" cursor="move" fill="#777" fill-opacity="0.3" stroke="#fff" shape-rendering="crispEdges" style="display: none;"></rect>
              <rect class="handle handle--w" cursor="ew-resize" style="display: none;"></rect>
              <rect class="handle handle--e" cursor="ew-resize" style="display: none;"></rect>
            </g>
          </g>
        </svg>
      </div>
      <div class="plot-legend">
        <div class="legend-item">
          <svg width="64" height="32">
            <g>
              <g class="theor-graph refTheor">
                <g>
                  <path d="M0,31L2,31L4,31L6,31L8,31L10,30.7L12,30.4L14,29.8L16,28.3L18,26.2L20,22.9L22,18.4L24,13.3L26,7.900000000000002L28,3.6999999999999993L30,1L32,1L34,3.400000000000002L36,7.900000000000002L38,13L40,18.4L42,22.6L44,26.2L46,28.3L48,29.5L50,30.4L52,30.7L54,31L56,31L58,31L60,31L62,31"></path>
                </g>
              </g>
            </g>
          </svg> <span class="legend-item-name">[C<sub>19</sub>H<sub>18</sub>N<sub>2</sub>O<sub>7</sub>S<sub>2</sub> - H]⁺<br> Theoretical</span></div>
        <div class="legend-item">
          <svg width="64" height="32">
            <g>
              <g class="sample-graph refSample">
                <g>
                  <circle cx="32" cy="4" r="4"></circle>
                </g>
                <g>
                  <line x1="32" x2="32" y1="4" y2="32"></line>
                </g>
                <g>
                  <rect width="48" height="28" x="8" y="4"></rect>
                </g>
              </g>
            </g>
          </svg> <span class="legend-item-name">[C<sub>19</sub>H<sub>18</sub>N<sub>2</sub>O<sub>7</sub>S<sub>2</sub> - H]⁺<br> Sample</span></div>
        <div class="legend-item">
          <svg width="64" height="32">
            <g>
              <g class="theor-graph compTheor">
                <g>
                  <path d="M0,31L2,31L4,31L6,31L8,31L10,30.7L12,30.4L14,29.8L16,28.3L18,26.2L20,22.9L22,18.4L24,13.3L26,7.900000000000002L28,3.6999999999999993L30,1L32,1L34,3.400000000000002L36,7.900000000000002L38,13L40,18.4L42,22.6L44,26.2L46,28.3L48,29.5L50,30.4L52,30.7L54,31L56,31L58,31L60,31L62,31"></path>
                </g>
              </g>
            </g>
          </svg> <span class="legend-item-name">[C<sub>8</sub>H<sub>10</sub>S<sub>2</sub> - H]⁻<br> Theoretical</span></div>
        <div class="legend-item">
          <svg width="64" height="32">
            <g>
              <g class="sample-graph compSample">
                <g>
                  <circle cx="32" cy="4" r="4"></circle>
                </g>
                <g>
                  <line x1="32" x2="32" y1="4" y2="32"></line>
                </g>
                <g>
                  <rect width="48" height="28" x="8" y="4"></rect>
                </g>
              </g>
            </g>
          </svg> <span class="legend-item-name">[C<sub>8</sub>H<sub>10</sub>S<sub>2</sub> - H]⁻<br> Sample</span></div>
      </div>
    </div>
  </div>
</div>
</div>
`;
